File: app/src/main/java/com/example/buildmypc/ui/parts/parts/PSU.java
package com.example.buildmypc.ui.parts.parts;

import android.os.Parcel;
import android.os.Parcelable;

import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;

public class PSU extends Part implements Parcelable { // stands for "Power Supply Unit"
	public static final Creator<PSU> CREATOR = new Creator<PSU>() {
		@NotNull
		@Contract("_ -> new")
		@Override
		public PSU createFromParcel(Parcel in) {
			return new PSU(in);
		}

		@NotNull
		@Contract(value = "_ -> new", pure = true)
		@Override
		public PSU[] newArray(int size) {
			return new PSU[size];
		}
	};
	private ArrayList<CountedString> connectorList;
	private String efficiencyRating;
	private boolean hasFan;
	private String formFactor;
	private int length; // in millimeters
	private String modular;
	private int wattage;

	public PSU(String model, String manufacturer) {
		super(model, manufacturer);
	}

	public PSU(@Nullable String model, @Nullable String manufacturer, ArrayList<CountedString> connectorList, String efficiencyRating, boolean hasFan, String formFactor, int length, String modular, int wattage) {
		super(model, manufacturer);
		this.connectorList = connectorList;
		this.efficiencyRating = efficiencyRating;
		this.hasFan = hasFan;
		this.formFactor = formFactor;
		this.length = length;
		this.modular = modular;
		this.wattage = wattage;
	}

	public PSU(Parcel in) {
		super(in.readString(), in.readString());
		in.readParcelableList(connectorList, connectorList.getClass().getClassLoader());
		efficiencyRating = in.readString();
		hasFan = in.readBoolean();
		formFactor = in.readString();
		length = in.readInt();
		modular = in.readString();
		wattage = in.readInt();
	}

	public ArrayList<CountedString> getConnectorList() {
		return connectorList;
	}

	public void setConnectorList(ArrayList<CountedString> connectorList) {
		this.connectorList = connectorList;
	}

	public String getEfficiencyRating() {
		return efficiencyRating;
	}

	public void setEfficiencyRating(String efficiencyRating) {
		this.efficiencyRating = efficiencyRating;
	}

	public boolean isHasFan() {
		return hasFan;
	}

	public void setHasFan(boolean hasFan) {
		this.hasFan = hasFan;
	}

	public String getFormFactor() {
		return formFactor;
	}

	public void setFormFactor(String formFactor) {
		this.formFactor = formFactor;
	}

	public int getLength() {
		return length;
	}

	public void setLength(int length) {
		this.length = length;
	}

	public String getModular() {
		return modular;
	}

	public void setModular(String modular) {
		this.modular = modular;
	}

	public int getWattage() {
		return wattage;
	}

	public void setWattage(int wattage) {
		this.wattage = wattage;
	}

	@Override
	public boolean equals(Object o) {
		if (this == o) return true;
		if (o == null || getClass() != o.getClass()) return false;
		if (!super.equals(o)) return false;

		PSU psu = (PSU) o;

		if (isHasFan() != psu.isHasFan()) return false;
		if (getLength() != psu.getLength()) return false;
		if (getWattage() != psu.getWattage()) return false;
		if (getConnectorList() != null ? !getConnectorList().equals(psu.getConnectorList()) : psu.getConnectorList() != null)
			return false;
		if (getEfficiencyRating() != null ? !getEfficiencyRating().equals(psu.getEfficiencyRating()) : psu.getEfficiencyRating() != null)
			return false;
		if (getFormFactor() != null ? !getFormFactor().equals(psu.getFormFactor()) : psu.getFormFactor() != null)
			return false;
		return getModular() != null ? getModular().equals(psu.getModular()) : psu.getModular() == null;
	}

	@Override
	public int hashCode() {
		int result = super.hashCode();
		result = 31 * result + (getConnectorList() != null ? getConnectorList().hashCode() : 0);
		result = 31 * result + (getEfficiencyRating() != null ? getEfficiencyRating().hashCode() : 0);
		result = 31 * result + (isHasFan() ? 1 : 0);
		result = 31 * result + (getFormFactor() != null ? getFormFactor().hashCode() : 0);
		result = 31 * result + getLength();
		result = 31 * result + (getModular() != null ? getModular().hashCode() : 0);
		result = 31 * result + getWattage();
		return result;
	}

	@Override
	public void writeToParcel(@NotNull Parcel dest, int flags) {
		super.writeToParcel(dest, flags);
		dest.writeParcelableList(connectorList, flags);
		dest.writeString(efficiencyRating);
		dest.writeBoolean(hasFan);
		dest.writeString(formFactor);
		dest.writeInt(length);
		dest.writeString(modular);
		dest.writeInt(wattage);
	}
}

File: app/src/main/java/com/example/buildmypc/ui/parts/parts/PSU.kt
package com.example.buildmypc.ui.parts.parts

import android.os.Parcel
import android.os.Parcelable
import android.os.Parcelable.Creator
import org.jetbrains.annotations.Contract
import java.util.*

class PSU : Part, Parcelable {
	var connectorList: ArrayList<CountedString?>? = null
	var efficiencyRating: String? = null
	var isHasFan = false
	var formFactor: String? = null
	var length // in millimeters
			= 0
	var modular: String? = null
	var wattage = 0

	constructor(model: String?, manufacturer: String?) : super(model, manufacturer)
	constructor(
		model: String?,
		manufacturer: String?,
		connectorList: ArrayList<CountedString?>?,
		efficiencyRating: String?,
		hasFan: Boolean,
		formFactor: String?,
		length: Int,
		modular: String?,
		wattage: Int
	) : super(model, manufacturer) {
		this.connectorList = connectorList
		this.efficiencyRating = efficiencyRating
		this.isHasFan = hasFan
		this.formFactor = formFactor
		this.length = length
		this.modular = modular
		this.wattage = wattage
	}

	constructor(`in`: Parcel) : super(`in`.readString(), `in`.readString()) {
		connectorList = `in`.createTypedArrayList(CountedString.CREATOR)
		efficiencyRating = `in`.readString()
		isHasFan = `in`.readBoolean()
		formFactor = `in`.readString()
		length = `in`.readInt()
		modular = `in`.readString()
		wattage = `in`.readInt()
	}

	override fun equals(other: Any?): Boolean {
		if (this === other) return true
		if (other == null || javaClass != other.javaClass) return false
		if (!super.equals(other)) return false
		val psu = other as PSU
		if (isHasFan != psu.isHasFan) return false
		if (length != psu.length) return false
		if (wattage != psu.wattage) return false
		if (if (connectorList != null) connectorList != psu.connectorList else psu.connectorList != null) return false
		if (if (efficiencyRating != null) efficiencyRating != psu.efficiencyRating else psu.efficiencyRating != null) return false
		if (if (formFactor != null) formFactor != psu.formFactor else psu.formFactor != null) return false
		return if (modular != null) modular == psu.modular else psu.modular == null
	}

	override fun hashCode(): Int {
		var result = super.hashCode()
		result = 31 * result + if (connectorList != null) connectorList.hashCode() else 0
		result = 31 * result + if (efficiencyRating != null) efficiencyRating.hashCode() else 0
		result = 31 * result + if (isHasFan) 1 else 0
		result = 31 * result + if (formFactor != null) formFactor.hashCode() else 0
		result = 31 * result + length
		result = 31 * result + if (modular != null) modular.hashCode() else 0
		result = 31 * result + wattage
		return result
	}

	override fun writeToParcel(dest: Parcel, flags: Int) {
		super.writeToParcel(dest, flags)
		dest.writeParcelableList(connectorList, flags)
		dest.writeString(efficiencyRating)
		dest.writeBoolean(isHasFan)
		dest.writeString(formFactor)
		dest.writeInt(length)
		dest.writeString(modular)
		dest.writeInt(wattage)
	}

	override val paramCount: Int
		get() = 9

	companion object {
		// stands for "Power Supply Unit"
		@JvmField
		val CREATOR: Creator<PSU> = object : Creator<PSU> {
			@Contract("_ -> new")
			override fun createFromParcel(`in`: Parcel) = PSU(`in`)

			@Contract(value = "_ -> new", pure = true)
			override fun newArray(size: Int): Array<PSU?> = arrayOfNulls(size)
		}
	}
}